src/main/java/com/fasterxml/jackson/databind/deser/std/StdValueInstantiator.java

```diff
@@ -459,9 +459,30 @@ public Object createFromBigDecimal(DeserializationContext ctxt, BigDecimal value
             }
         }
 
+        // 13-Dec-2020, ckozak: Unlike other types, BigDecimal values may be represented
+        // with less precision as doubles. When written to a TokenBuffer for polymorphic
+        // deserialization the most specific type is recorded, though a less precise
+        // floating point value may be needed.
+        if(_fromDoubleCreator != null && isValidDouble(value)) {
+            Object arg = value.doubleValue();
+            try {
+                return _fromDoubleCreator.call1(arg);
+            } catch (Throwable t0) {
+                return ctxt.handleInstantiationProblem(_fromDoubleCreator.getDeclaringClass(),
+                        arg, rewrapCtorProblem(ctxt, t0));
+            }
+        }
+
         return super.createFromBigDecimal(ctxt, value);
     }
 
+    // BigDecimal cannot represent special values NaN, positive infinity, or negative infinity.
+    // When the value cannot be represented as a double, positive or negative infinity is returned.
+    static boolean isValidDouble(BigDecimal value) {
```

Comment:
Maybe "canConvertToDouble()" would be even more accurate name (following what `JsonNode` uses)?

Reply:
I can change this after merge. And also figure out how to cleanly avoid double-cast, just because that is potentially an expensive operation I think (from 10- to 2-base floating number conversion is probably the most expensive part, similar to parsing from textual to 2-based)

Reply:
Ya, I went back and forth on that -- I ended up double-casting in order to provide a clean utility method that I could add test coverage for.